File: src/monitor_factory.cpp
#include "monitor_factory.h"

namespace monitoring
{
    /**
     * Returns a pointer to a hardware monitor specified by the given name
     *
     * @param monitor_name monitor description name as specified in 'config_enums/HardwareMonitorNames'
     */
    std::shared_ptr<MonitorBase> MonitorFactory::getHardwareMonitor(config::MonitorModeConfig monitor_config_params)
    {
        if (monitor_config_params.name == HardwareMonitorNames::LASER_DEVICE_MONITOR)
        {
            std::shared_ptr<LaserDeviceMonitor> monitor(new LaserDeviceMonitor(monitor_config_params));
            return monitor;
        }
        // else if (monitor_name == HardwareMonitorNames::LASER_FUNCTIONAL_MONITOR)
        // {
        //     std::shared_ptr<LaserFunctionalMonitor> monitor(new LaserFunctionalMonitor());
        //     return monitor;
        // }
        // else if (monitor_name == HardwareMonitorNames::LASER_HEARTBEAT_MONITOR)
        // {
        //     std::shared_ptr<LaserHeartbeatMonitor> monitor(new LaserHeartbeatMonitor());
        //     return monitor;
        // }
    }

    /**
     * Returns a pointer to a software monitor specified by the given name
     *
     * @param monitor_name monitor description name as specified in 'config_enums/SoftwareMonitorNames'
     */
    std::shared_ptr<MonitorBase> MonitorFactory::getSoftwareMonitor(config::MonitorModeConfig monitor_config_params)
    {
        return NULL;
    }
}

File: component_monitoring/monitor_factory.py
import importlib
from component_monitoring.monitor_base import MonitorBase

'''A factory for creating component monitors

@author Alex Mitrevski, Santosh Thoduka
@contact aleksandar.mitrevski@h-brs.de, santosh.thoduka@h-brs.de
'''
class MonitorFactory(object):
    '''Returns a hardware monitor as specified by the given name

    Keyword arguments:
    @param monitor_name monitor description name as specified in 'config_enums/HardwareMonitorNames'

    '''
    @staticmethod
    def get_hardware_monitor(component_name, monitor_config_params):
        try:
            monitor_name = monitor_config_params.name
            module_name = 'component_monitoring.monitors.hardware.' \
                          + component_name + '.' + monitor_name
            class_name = ''.join(x.title() for x in monitor_name.split('_'))
            MonitorClass = getattr(importlib.import_module(module_name),
                                   class_name)
            return MonitorClass(monitor_config_params)
        except Exception as e:
            print("While initialising", component_name, "got exception", str(e))
            return MonitorBase(monitor_config_params)

    '''Returns a software monitor as specified by the given name

    Keyword arguments:
    @param monitor_name monitor description name as specified in 'config_enums/HardwareMonitorNames'

    '''
    @staticmethod
    def get_software_monitor(component_name, monitor_config_params):
        try:
            monitor_name = monitor_config_params.name
            module_name = 'component_monitoring.monitors.software.' \
                          + component_name + '.' + monitor_name
            class_name = ''.join(x.title() for x in monitor_name.split('_'))
            MonitorClass = getattr(importlib.import_module(module_name),
                                   class_name)
            return MonitorClass(monitor_config_params)
        except Exception as e:
            print("While initialising", component_name, "got exception", str(e))
            return MonitorBase(monitor_config_params)
